src/components/sketches/Scene4.ts

```diff
@@ -0,0 +1,170 @@
+          p.rotate(v.heading());
+          const endpoint = p.abs(m) * SCL;
+          p.line(0, 0, endpoint, 0);
+          if (m < 0) {
+            p.stroke('red');
+          } else {
+            p.stroke('green');
+          }
+          p.line(endpoint - 2, 0, endpoint, 0);
+          p.pop();
+        }
+        flowfield[index] = v;
+        xoff += INC;
+      }
+      yoff += INC;
+    }
+    magOff += MAG_INC;
+    zoff += INC_START;
+    start -= MAG_INC;
+
+    if (!showField) {
+      for (let i = 0; i < particles.length; i++) {
+        particles[i].follow(flowfield);
+        particles[i].update();
+        particles[i].edges();
+        particles[i].show();
+      }
+
+      if (p.random(10) > 5 && particles.length < 2500) {
+        const rnd = p.floor(p.noise(zoff) * 20);
+        for (let i = 0; i < rnd; i++) {
+          particles.push(new Particle());
+        }
+      } else if (particles.length > 2000) {
+        const rnd = p.floor(p.random(10));
+        for (let i = 0; i < rnd; i++) {
+          particles.shift();
+        }
+      }
+    }
+  };
+
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, false);
+  };
+
+  class Particle {
+    public pos: p5.Vector;
+    public vel: p5.Vector;
+    private _acc: p5.Vector;
+    private _maxSpeed: number;
+    public _prevPos: p5.Vector;
+    private _bubble: number;
+
+    constructor() {
+      this.pos = p.createVector(p.random(window.innerWidth),p.random(window.innerHeight));
+      this.vel = p.createVector(0, 0);
+      this._acc = p.createVector(0, 0);
+      this._maxSpeed = 2;
+      this._bubble = p.random();
+
+      this._prevPos = this.pos.copy();
+    }
+
+    public update() {
+      this.vel.add(this._acc);
+      this.vel.limit(this._maxSpeed);
+      this.pos.add(this.vel);
+      this._acc.mult(0);
+    }
+
+    public show() {
+      p.strokeWeight(STROKE_WEIGHT);
+      p.line(this.pos.x, this.pos.y, this._prevPos.x, this._prevPos.y);
+      this.updatePrev();
+      //point(this.pos.x, this.pos.y);
+    }
+
+    public updatePrev() {
+      this._prevPos.x = this.pos.x;
+      this._prevPos.y = this.pos.y;
+    }
+
+    public edges() {
+      this.updatePrev();
+    }
+
+    public follow(vectors: p5.Vector[]) {
+      const x = p.floor(this.pos.x / SCL);
+      const y = p.floor(this.pos.y / SCL);
+      const index = x + y * cols;
+      this._acc.add(vectors[index]);
+
+      /** drawing depending on bubble */
+      const di = p.dist(p.mouseX, p.mouseY, this.pos.x, this.pos.y);
+      if (di <= TRANSPARENCY_RADIUS) {
+        p.stroke(p.color(77, 184, 165));
+        if (this._bubble < BUBBLE_FREQUENCY) {
+          if (p.random() < BUBBLE_FREQUENCY_ON_LINES) {
+            p.fill(77, 184, 165);
+            p.ellipse(this.pos.x, this.pos.y, p.width*.006);
+          }
+        }
+      } else {
+        p.stroke(`rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
+      }
```

Comment:
```suggestion
      p.stroke(di <= TRANSPARENCY_RADIUS ? p.color(77, 184, 165) : `rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
      if (this._bubble < BUBBLE_FREQUENCY && p.random() < BUBBLE_FREQUENCY_ON_LINES) {
        p.fill(77, 184, 165);
        p.ellipse(this.pos.x, this.pos.y, p.width*.006);
      }
```

Reply:
this suggestion makes bubbles every where, but we only want it on hover area within `di`

Reply:
okay then just modify to make it work

```ts
      p.stroke(di <= TRANSPARENCY_RADIUS ? p.color(77, 184, 165) : `rgba(${41}, ${157}, ${116}, ${TRANSPARENCY})`);
      if (di <= TRANSPARENCY_RADIUS && this._bubble < BUBBLE_FREQUENCY && p.random() < BUBBLE_FREQUENCY_ON_LINES) {
        p.fill(77, 184, 165);
        p.ellipse(this.pos.x, this.pos.y, p.width*.006);
      }
```

Reply:
yes that works out